reuse.go

```diff
@@ -0,0 +1,140 @@
+package libp2pquic
+
+import (
+	"net"
+	"sync"
+	"sync/atomic"
+
+	"github.com/vishvananda/netlink"
+)
+
+type reuseConn struct {
+	net.PacketConn
+	refCount int32 // to be used as an atomic
+}
+
+func newReuseConn(conn net.PacketConn) *reuseConn {
+	return &reuseConn{PacketConn: conn}
+}
+
+func (c *reuseConn) IncreaseCount() { atomic.AddInt32(&c.refCount, 1) }
+func (c *reuseConn) DecreaseCount() { atomic.AddInt32(&c.refCount, -1) }
+func (c *reuseConn) GetCount() int  { return int(atomic.LoadInt32(&c.refCount)) }
+
+type reuse struct {
+	mutex sync.Mutex
+
+	unicast map[string] /* IP.String() */ map[int] /* port */ *reuseConn
+	// global contains connections that are listening on 0.0.0.0 / ::
+	global map[int]*reuseConn
+}
+
+func newReuse() *reuse {
+	return &reuse{
+		unicast: make(map[string]map[int]*reuseConn),
+		global:  make(map[int]*reuseConn),
+	}
+}
+
+func (r *reuse) getSourceIPs(network string, raddr *net.UDPAddr) ([]net.IP, error) {
+	// Determine the source address that the kernel would use for this IP address.
+	// Note: This only works on Linux.
+	// On other OSes, this will return a netlink.ErrNotImplemetned.
+	routes, err := (&netlink.Handle{}).RouteGet(raddr.IP)
+	if err != nil {
+		return nil, err
+	}
+
+	ips := make([]net.IP, 0, len(routes))
+	for _, route := range routes {
+		ips = append(ips, route.Src)
+	}
+	return ips, nil
+}
+
+func (r *reuse) Dial(network string, raddr *net.UDPAddr) (*reuseConn, error) {
+	conn, err := r.dial(network, raddr)
+	if err != nil {
+		return nil, err
+	}
+	conn.IncreaseCount()
```

Comment:
Race?

Reply:
We need to check when incrementing to see if the count dropped below 0.

(really, closing should set the count to -max).

Reply:
Right, we need to do this while holding the mutex. Fixed.

Reply:
We're racing against a separate (theoretical) garbage collection event. I guess we can have that hold the lock but it may be better to just _poison_ the connection such that `IncreaseCount` returns an error (and we try again).